npm/networkPolicyController.go

```diff
@@ -0,0 +1,491 @@
+		npMgr:                  npMgr,
+		isAzureNpmChainCreated: false,
+	}
+
+	npInformer.Informer().AddEventHandler(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    netPolController.addNetPol,
+			UpdateFunc: netPolController.updateNetPol,
+			DeleteFunc: netPolController.deleteNetPol,
+		},
+	)
+	return netPolController
+}
+
+// needSync checks whether the obj is valid network policy object.
+func (c *networkPolicyController) needSync(obj interface{}) (string, error) {
+	var key string
+	_, ok := obj.(*networkingv1.NetworkPolicy)
+	if !ok {
+		return key, fmt.Errorf("cannot cast obj (%v) to network policy obj", obj)
+	}
+
+	var err error
+	if key, err = cache.MetaNamespaceKeyFunc(obj); err != nil {
+		return key, fmt.Errorf("error due to %s", err)
+	}
+
+	return key, nil
+}
+
+func (c *networkPolicyController) addNetPol(obj interface{}) {
+	key, err := c.needSync(obj)
+	if err != nil {
+		utilruntime.HandleError(err)
+		return
+	}
+
+	c.workqueue.Add(key)
+}
+
+func (c *networkPolicyController) updateNetPol(old, new interface{}) {
+	key, err := c.needSync(new)
+	if err != nil {
+		utilruntime.HandleError(err)
+		return
+	}
+
+	// needSync already checked validation of casting new network policy.
+	newNetPol, _ := new.(*networkingv1.NetworkPolicy)
+	oldNetPol, ok := old.(*networkingv1.NetworkPolicy)
+	if ok {
+		if oldNetPol.ResourceVersion == newNetPol.ResourceVersion {
+			// Periodic resync will send update events for all known network plicies.
+			// Two different versions of the same network policy will always have different RVs.
+			return
+		}
+	}
+
+	c.workqueue.Add(key)
+}
+
+func (c *networkPolicyController) deleteNetPol(obj interface{}) {
+	_, ok := obj.(*networkingv1.NetworkPolicy)
+	// DeleteFunc gets the final state of the resource (if it is known).
+	// Otherwise, it gets an object of type DeletedFinalStateUnknown.
+	// This can happen if the watch is closed and misses the delete event and
+	// the controller doesn't notice the deletion until the subsequent re-list
+	if !ok {
+		tombstone, ok := obj.(cache.DeletedFinalStateUnknown)
+		if !ok {
+			metrics.SendErrorLogAndMetric(util.NetpolID, "[NETPOL DELETE EVENT] Received unexpected object type: %v", obj)
+			utilruntime.HandleError(fmt.Errorf("error decoding object, invalid type"))
+			return
+		}
+
+		if _, ok = tombstone.Obj.(*networkingv1.NetworkPolicy); !ok {
+			metrics.SendErrorLogAndMetric(util.NetpolID, "[NETPOL DELETE EVENT] Received unexpected object type (error decoding object tombstone, invalid type): %v", obj)
+			utilruntime.HandleError(fmt.Errorf("error decoding object tombstone, invalid type"))
+			return
+		}
+	}
+
+	var key string
+	var err error
+	if key, err = cache.MetaNamespaceKeyFunc(obj); err != nil {
+		utilruntime.HandleError(err)
+		return
+	}
+
+	netPolCachedKey := util.GetNSNameWithPrefix(key)
+
+	// (TODO): Reduce scope of lock later
+	c.npMgr.Lock()
+	defer c.npMgr.Unlock()
```

Comment:
We can optimize this lock from defer to explicitly mentioning below RawNpMap exists check. thta way we are not locking when Add is being performed. Probably we can expand this to other controllers later.

Reply:
Good point. Another important thing is we need to decouple all shared maps (e.g., RawNpMap, etc) from npm.go to each controller and then carefully walk through them again. Even we may encapsulate each map with lock as a struct. Let's understand each maps in details and start this process as another PR.